packages/tunnel-client/src/lib/tunnel-http2-cluster.ts

```diff
@@ -0,0 +1,146 @@
+  private readonly opts: TunnelClusterOpts;
+  private readonly sockets: net.Socket[];
+  private readonly server: Http2Server;
+
+  constructor(opts: TunnelHTTP2ClusterOpts) {
+    super();
+    this.logger = opts.logger;
+    this.opts = opts;
+
+    this.sockets = opts.sockets;
+
+    this.server = createServer({
+      settings: {
+        initialWindowSize: HTTP2_WINDOW_SIZE,
+      },
+      maxSessionMemory: HTTP2_MAX_SESSION_MEMORY,
+    });
+
+    this.server.on("session", (session) => {
+      session.once("remoteSettings", () => {
+        session.setLocalWindowSize(HTTP2_WINDOW_SIZE);
+      });
+    });
+  }
+
+  startListening() {
+    const opt = this.opts;
+
+    const localHost = opt.localHost;
+    const localPort = opt.localPort;
+    const localProtocol = opt.localHttps ? "https" : "http";
+
+    const allowInvalidCert = opt.allowInvalidCert;
+    const localCertOpts =
+      localProtocol === "http" || allowInvalidCert
+        ? { rejectUnauthorized: false }
+        : {
+            cert: readFileSync(opt.localCert as string),
+            key: readFileSync(opt.localKey as string),
+            ca: opt.localCa ? [readFileSync(opt.localCa)] : undefined,
+          };
+
+    const agent = new Agent();
+
+    this.server.on("request", (req, res) => {
+      this.emit("request", {
+        method: req.method ?? "unknown",
+        path: req.url ?? "unknown",
+      });
+
+      // Drop host & connection header from the original request. Let Node handle it.
+      const { host, connection, ..._headersToForward } = req.headers;
+
+      // Also drop HTTP2 pseudo headers
+      const headersToForward = Object.keys(_headersToForward).reduce(
+        (acc, key) => {
+          if (!key.startsWith(":")) {
+            acc[key] = _headersToForward[key];
+          }
+          return acc;
+        },
+        {} as Record<string, string | string[] | undefined>
+      );
+
+      // Forward the request to the local server
+      const clientReq = request(
+        {
+          agent,
+          host: localHost,
+          port: localPort,
+          path: req.url,
+          method: req.method,
+          headers: headersToForward,
+          ...localCertOpts,
+        },
+        (clientRes) => {
+          // Drop HTTP1 specific headers
+          const {
+            connection,
+            "keep-alive": _,
+            "transfer-encoding": __,
+            ...headersToForward
+          } = clientRes.headers;
+
+          res.writeHead(clientRes.statusCode as number, headersToForward);
+
+          pipeline(clientRes, res, (err) => {
+            if (err) {
+              this.logger.error("Response pipeline error", err);
+            }
+          });
+        }
+      );
+
+      pipeline(req, clientReq, (err) => {
+        if (err) {
+          this.logger.error("Request pipeline error", err);
+        }
+      });
+    });
+
+    this.sockets.forEach((socket) => {
+      this.server.emit("connection", socket);
```

Comment:
Is this "connection" event used anywhere? I can't work out how these sockets are used...